consensus/hotstuff/integration/instance_test.go

```diff
@@ -434,3 +448,23 @@ func (in *Instance) Run() error {
 
 	}
 }
+
+func (in *Instance) ProcessBlock(proposal *model.Proposal) {
+	in.updatingBlocks.RLock()
+	_, parentExists := in.headers[proposal.Block.QC.BlockID]
+	defer in.updatingBlocks.RUnlock()
+
+	if parentExists {
+		next := proposal
+		for next != nil {
+			in.headers[next.Block.BlockID] = model.ProposalToFlow(next)
```

Comment:
⚠️  We are writing to `in.headers` but have only acquired the read lock of `updatingBlocks`.